block-layer/codecs/dag-pb.md

```diff
@@ -0,0 +1,59 @@
+# DagPB Spec
+
+DagPB does not support the full ["IPLD Data Model."](../../data-model-layer/data-model.md)
+
+## Format
+
+The DagPB IPLD format is a legacy format implemented with a single protobuf.
+
+```protobuf
+// An IPFS MerkleDAG Link
+message PBLink {
+
+  // binary CID (with no multibase prefix) of the target object
```

Comment:
Is this correct?  You can have v1 CIDs as DAGLinks which can contain a multibase prefix.

Reply:
I haven't addressed this as I also don't know (@Stebalien probably does).

Reply:
https://github.com/ipld/js-ipld-dag-pb/blob/c4b83cbb69821ca6da12190105b0e78d27688630/src/serialize.js#L23
https://github.com/ipfs/go-merkledag/blob/2c1a891f6f1b2bbb6442b9da62e5ece591fbdad9/pb/merkledag.pb.go#L425

these both looks like raw CID byte arrays

Reply:
CIDs don't contain a multibase prefix.

I created a script to verify this (yes, I am using a sledgehammer to crack a nut, but it gave me the chance to finally use the new js-ipld-dag API as well as playing a bit with go-ipld). The output is:

```console
$ node --experimental-modules combined.js
(node:985) ExperimentalWarning: The ESM module loader is experimental.
actual hash:   89b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8
cidv0:     122089b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8
cidv1: 0170122089b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8
```

And here's the script in case someone wants to run it locally:

```js
import dagPb from 'ipld-dag-pb'
const { DAGLink, DAGNode} = dagPb
import Pbf from 'pbf'

// Creating the data

const node = new DAGNode('some data')

// base58btc - cidv0 - dag-pb - sha2-256-256-89b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8
const cidv0 = 'QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ'
const linkv0 = new DAGLink('cidv0link', 10, cidv0)
node.addLink(linkv0)

// base32 - cidv1 - dag-pb - sha2-256-256-89b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8
const cidv1 = 'bafybeiejwhmzk7zep3qpxktaojv6jyjy72myhrmrtxwbug4kipt62zijza'
const linkv1 = new DAGLink('cidv1link', 11, cidv1)
node.addLink(linkv1)

//console.log(node)
const serialized = node.serialize()
//console.log(serialized)


// Reading the data (generated by pbf)

// PBLink ========================================

const PBLink = {};

PBLink.read = function (pbf, end) {
    return pbf.readFields(PBLink._readField, {Hash: null, Name: "", Tsize: 0}, end);
};
PBLink._readField = function (tag, obj, pbf) {
    if (tag === 1) obj.Hash = pbf.readBytes();
    else if (tag === 2) obj.Name = pbf.readString();
    else if (tag === 3) obj.Tsize = pbf.readVarint();
};

// PBNode ========================================

const PBNode = {};

PBNode.read = function (pbf, end) {
    return pbf.readFields(PBNode._readField, {Links: [], Data: null}, end);
};
PBNode._readField = function (tag, obj, pbf) {
    if (tag === 2) obj.Links.push(PBLink.read(pbf, pbf.readVarint() + pbf.pos));
    else if (tag === 1) obj.Data = pbf.readBytes();
};


const pbf = new Pbf(serialized)
const data = PBNode.read(pbf)
console.log('actual hash:   89b1d9957f247ee0fbaa60726be4e138fe9983c5919dec1a1b8a43e7ed6509c8')
const cidv0Read = data.Links[0].Hash;
console.log('cidv0:    ', cidv0Read.toString('hex'))
const cidv1Read = data.Links[1].Hash;
console.log('cidv1:', cidv1Read.toString('hex'))
```

Reply:
https://github.com/multiformats/cid#how-does-it-work

```
<multibase-prefix><cid-version><multicodec-content-type><multihash-content-address>
```

[CIDv0](http://127.0.0.1:8080/ipns/cid.ipfs.io/#QmXc9raDM1M5G5fpBnVyQ71vR4gbnskwnB9iMEzBuLgvoZ) does not contain a multibase prefix as it's assumed to be `base58btc`, [CIDv1](bafybeiejwhmzk7zep3qpxktaojv6jyjy72myhrmrtxwbug4kipt62zijza) does.

Reply:
@achingbrain yes, and it also mentions there:

> NOTE: Binary (not text-based) protocols and formats may omit the multibase prefix when the encoding is unambiguous.

And this is what dag-pb is doing. The CID (no matter which one) is stored in binary without the multibase prefix (as the code above shows).

Reply:
Good point, well presented.

Reply:
> Since that's the case, you could probably remove the `(with no multibase prefix)` altogether.

It's not strictly needed, but I think it adds clarity.